Sources/AsyncAlgorithms/Retry/Backoff.swift

```diff
@@ -0,0 +1,347 @@
+#if compiler(>=6.2)
+/// A protocol for defining backoff strategies that generate delays between retry attempts.
+///
+/// Each call to `nextDuration()` returns the delay for the next retry attempt. Strategies are
+/// naturally stateful. For instance, they may track the number of invocations or the previously
+/// returned duration to calculate the next delay.
+///
+/// - Precondition: Strategies should only increase or stay the same over time, never decrease.
+///   Decreasing delays may cause issues with modifiers like jitter which expect non-decreasing values.
+///
+/// ## Example
+///
+/// ```swift
+/// var strategy = Backoff.exponential(factor: 2, initial: .milliseconds(100))
+/// strategy.nextDuration() // 100ms
+/// strategy.nextDuration() // 200ms
+/// strategy.nextDuration() // 400ms
+/// ```
+@available(iOS 16.0, macCatalyst 16.0, macOS 13.0, tvOS 16.0, visionOS 1.0, watchOS 9.0, *)
+public protocol BackoffStrategy<Duration> {
+  associatedtype Duration: DurationProtocol
+  mutating func nextDuration() -> Duration
+}
+
+@available(iOS 16.0, macCatalyst 16.0, macOS 13.0, tvOS 16.0, visionOS 1.0, watchOS 9.0, *)
+@usableFromInline struct ConstantBackoffStrategy<Duration: DurationProtocol>: BackoffStrategy {
+  @usableFromInline let constant: Duration
+  @usableFromInline init(constant: Duration) {
+    precondition(constant >= .zero, "Constant must be greater than or equal to 0")
+    self.constant = constant
+  }
+  @inlinable func nextDuration() -> Duration {
+    return constant
+  }
+}
+
+@available(iOS 16.0, macCatalyst 16.0, macOS 13.0, tvOS 16.0, visionOS 1.0, watchOS 9.0, *)
+@usableFromInline struct LinearBackoffStrategy<Duration: DurationProtocol>: BackoffStrategy {
+  @usableFromInline var current: Duration
+  @usableFromInline let increment: Duration
+  @usableFromInline init(increment: Duration, initial: Duration) {
+    precondition(initial >= .zero, "Initial must be greater than or equal to 0")
+    precondition(increment >= .zero, "Increment must be greater than or equal to 0")
+    self.current = initial
+    self.increment = increment
+  }
+  @inlinable mutating func nextDuration() -> Duration {
+    defer { current += increment }
+    return current
+  }
+}
+
+@available(iOS 16.0, macCatalyst 16.0, macOS 13.0, tvOS 16.0, visionOS 1.0, watchOS 9.0, *)
+@usableFromInline struct ExponentialBackoffStrategy<Duration: DurationProtocol>: BackoffStrategy {
+  @usableFromInline var current: Duration
+  @usableFromInline let factor: Int
+  @usableFromInline init(factor: Int, initial: Duration) {
+    precondition(initial >= .zero, "Initial must be greater than or equal to 0")
+    self.current = initial
+    self.factor = factor
+  }
+  @inlinable mutating func nextDuration() -> Duration {
+    defer { current *= factor }
```

Comment:
Looks like this multiplication can overflow when retrying many times. I just saw it happen with `.exponential(factor: 2, initial: .seconds(5))`. This should check for overflows and cap the value to the max of the duration type.

Reply:
@bobergj I've updated the implementation and proposal: https://forums.swift.org/t/pitch-retry-backoff/82483/2. Thanks again.